File: solutions/binary-tree-inorder-traversal/binary-tree-inorder-traversal.py
// @Title: Binary Tree Inorder Traversal
// @Author: colinjxc
// @Date: 2022-02-04T13:13:20+08:00
// @URL: https://leetcode-cn.com/problems/binary-tree-inorder-traversal


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        stack = []
        res = []
        while root or len(stack) > 0:
            while root:
                stack.append(root)
                root = root.left
            root = stack.pop()
            res.append(root.val)
            root = root.right
        return res

File: solutions/binary-tree-inorder-traversal/binary-tree-inorder-traversal.go
// @Title: Binary Tree Inorder Traversal
// @Author: colinjxc
// @Date: 2021-07-07T11:54:48+08:00
// @URL: https://leetcode-cn.com/problems/binary-tree-inorder-traversal


/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func inorderTraversal(root *TreeNode) []int {
	res := []int{}
	var inorder func(*TreeNode)

	inorder = func(n *TreeNode) {
		if n == nil {
			return
		}
		inorder(n.Left)
		res = append(res, n.Val)
		inorder(n.Right)
	}

	inorder(root)
	return res
}

